# TaxOnTree: Including taxonomic information in phylogenetic trees

*Sakamoto T.* and *J. M. Ortega*

## 1. About TaxOnTree and this document

[TaxOnTree](http://biodados.icb.ufmg.br/taxontree) is a bioinformatics tool that embeds taxonomic information from [NCBI Taxonomy](https://www.ncbi.nlm.nih.gov/taxonomy) into phylogenetic trees. Trees generated by TaxOnTree allow users to easily access the taxonomic information of all entities comprising a tree. This could be performed either by colorizing the tree according to a taxonomic data or by adding/changing labels to evidence a taxonomic data on the tips, branches or nodes.

This document is a guide for users to understand TaxOnTree features and to explore all its resources,  making you an expert in taxonomy. 

## 2. TaxOnTree features

### 2.1. Phylogenetic pipeline

TaxOnTree has a simple phylogenetic pipeline implemented on it that allows several input format. Phylogenetic steps are schematized in Figure X.

### 2.2. Determining the lowest common ancestor (LCA)

LCA of two or more organisms represents the most recent ancestor that all organisms in a set have in common. Determining LCA is a way to measure the taxonomic relationship among the organisms in the analysis. 

To determine it, TaxOnTree takes advantage of the hierarchical structure from NCBI Taxonomy. To illustrate this, a short taxonomic lineages of human, dog and frog from NCBI Taxonomy are represented in the **Figure S2**. Walking through the human taxonomic lineage, beginning from the root (level 0), we could observe that several taxa comprising the human lineage are shared with the other two species. However, in different point of the human lineage, frog and dog lineages take different route. The frog lineage diverge from the human lineage after the level 17 (Amniota), while, in dog lineage, this occur at level 21 (Boreoeutheria). Those taxa, which are the last taxa shared between the pairs of lineages human X dog and human X frog, are what we denominate as LCA. The higher the LCA level, more recent was the first divergence between the species in comparison. So, for instance, by comparing the LCA level, we could claim that human is more closely related to dog than to frog.  

<img src="/img/lca_frog.png" width=650px/>

**Figure S2**: Determining LCA. Taxonomic lineage of human, dog and frog are shown. LCA between human and dog is Boreoeutheria, while LCA between human and frog is Amniota.

### 2.3. Missing ranks and Taxallnomy database

 Whenever we are querying for a taxonomic rank from NCBI Taxonomy, two issues have to be considered :
 * Some taxonomic ranks are absent in a taxonomic lineage, e.g. there is no taxon for subclass, superclass, and subphylum in the human lineage (**Figure S3A**);
 * Some taxa found in a taxonomic lineage do not have a taxonomic rank. These taxa are referred as *no rank*, e.g. Theria, Eutheria, Boroeutheria and others are taxa without rank in the human lineage (**Figure S3A**).  
  
 To handle these issues, we use [Taxallnomy](http://biodados.icb.ufmg.br/taxallnomy), a taxonomic database which provides a taxonomic lineage with all ranks for all taxa comprising the NCBI Taxonomy. Taxallnomy provides a balanced version of the taxonomic tree from NCBI Taxonomy which its hierarchical levels are correspondent to the taxonomic ranks. **Figure S3B** shows the human taxonomic lineage retrieved from Taxallnomy database. Ranks that are originally missing in the human taxonomic lineage are filled by this data.  
<img src="/img/taxontree_taxsimple.png" width=650px/>
  
**Figure S3**: Human taxonomic lineage from (A) NCBI Taxonomy and from (B) Taxallnomy. Taxa with a taxonomic rank assigned are in blue and taxa exclusive from Taxallnomy are in red.

## 3. Command-line version

TaxOnTree source code can be dowloaded in TaxOnTree SourceForge page. Here we describe how to install it in a local machine.

### 3.1. Prerequisites:

Some prerequisites are required to run TaxOnTree. They are:

* Unix Platform;
* Perl;
* Internet connection and/or TaxOnTree databases;
* Perl module IO::Socket::SSL (only for accessing info from NCBI via internet);
* [FigTree](http://tree.bio.ed.ac.uk/software/figtree/);
* Third-party software (only for phylogenetic pipeline).

#### 3.1.1 Unix Platform

TaxOnTree was tested on CentOS and on MacOS, but it should work on any Unix platforms.

#### 3.1.2. Perl

Almost all Unix platforms have Perl 5 installed. To verify that, type perl -v inside a command shell. If not, follow the instructions on [Perl website](https://www.perl.org/get.html) for its installation.

#### 3.1.3. Internet connection and/or TaxOnTree databases;

TaxOnTree can retrieve sequence and taxonomic information of accessions by two ways: via REST request from NCBI or Uniprot servers (require internet connection) and/or by consulting local databases. TaxOnTree preferentially retrieves information from local databases because it is certainly faster than requesting them from the web. However, if there is no local databases installed, or if there is local databases but some accessions are missing on them, TaxOnTree can retrieve these data from the NCBI or Uniprot servers, if the machine is connected to internet.  

There are two databases that TaxOnTree consults locally:

* MySQL database;
* Sequence database;

TaxOnTre MySQL database has relational tables that associate all accessions found in NCBI and Uniprot database to taxonomic information. This database can be downloaded in our [Sourceforge page]() and loaded to a local MySQL. The database provided in Sourceforge is monthly updated, but we also provide a [Perl script]() that downloads all necessary data from NCBI and Uniprot servers and generates the TaxOnTree MySQL tables. To enable TaxOnTree to consult the local MySQL, see [Configuring MySQL database]().

TaxOnTree also consults sequence databases to obtain sequence data that are required in some steps of the phylogenetic pipeline. These databases are BLAST-formatted sequence databases generated using *makeblastdb* from [Standalone-BLAST+ suite](https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastDocs&DOC_TYPE=Download). Sequence databases used in our web server can be downloaded in our [Sourceforge page](), but you can also make your own database using protein sequences from NCBI or Uniprot (See [Making your own sequence database]()).

> **Note**: TaxOnTree databases take up a lot of hard disk space (~30 GB). So, check your demand and evaluate if it is worth having a local database installed or if only web requests should be enough for your analysis.

#### 3.1.4. PERL module IO::Socket::SSL

Because NCBI server uses HTTPS to communicate, an additional Perl module ([IO::Socket::SSL](http://search.cpan.org/~sullr/IO-Socket-SSL-2.052/lib/IO/Socket/SSL.pod)) is necessary for TaxOnTree to request information to this server via REST. This module can be easily installed in your machine using [CPAN](http://www.cpan.org/modules/INSTALL.html) by typing the following commands in the terminal:

```bash
> cpan
cpan> install IO::Socket::SSL
cpan> exit
```

This should install IO::Socket::SSL module and all its dependencies.

#### 3.1.5. FigTree

[FigTree](http://tree.bio.ed.ac.uk/software/figtree/) is a free graphical viewer of phylogenetic trees developed in Java by Andrew Rambaut group. TaxOnTree output is made to be visualized in this software. In its website, there are versions for MacOS, Linux and Windows. Pick the one that is more convenient for you.

#### 3.1.6. Third-party software

*This requirement is only necessary if TaxOnTree phylogenetic pipeline is required. If your input is a tree file in Newick format, than no third-party software are necessary for your analysis.* 

The third-party software required by TaxOnTree are those software that will comprise the phylogenetic pipeline. Those software are divide in the following types:

* *Blast search* - this is performed exclusively by [Blast+](https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastDocs&DOC_TYPE=Download);
* *Sequence aligner* - software that perform sequence alignment. Must have Multi-FASTA file as input and as output, e.g. [MUSCLE](https://www.drive5.com/muscle/), [Clustal Omega](http://www.clustal.org/omega/), [Kalign](http://msa.sbc.su.se/cgi-bin/msa.cgi);
* *Sequence trimmer* - software that trim a sequence alignment according to its quality. Must have Multi-FASTA file as input and as output, e.g. [trimAl](http://trimal.cgenomics.org/);
* *Tree reconstructor* - software that reconstruct the phylogenetic history of a set of sequences. Must have an aligned Multi-FASTA as input and a tree in Newick format as output, e.g. [FastTree](http://www.microbesonline.org/fasttree/#Install).

Third-party software that met the conditions above can be incorporated in TaxOnTree phylogenetic pipeline. To make these software viewable by TaxOnTree, they have to be installed in your system or, alternatively, you can compile them and move their executables to the bin folder that accompanies TaxOnTree package. Furthermore, parameters and command lines that TaxOnTree executes of each third-party software have to be listed in CONFIG.xml file that accompanies TaxOnTree package. The original CONFIG.xml file has some third-party software with their default parameters configured. They are listed on table X.

Table X: Third-party software in CONFIG.xml.

| third-party software | executable name | phylogenetic pipeline step | link |
|-----------|-----------|-----------|-----------|
| Blast+ | blastp and blastdbcmd | blast search | https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastDocs&DOC_TYPE=Download |
| MUSCLE | muscle | sequence aligner | http://www.drive5.com/muscle/ |
| Clustal Omega | clustalo | sequence aligner | http://www.clustal.org/omega/ |
| Kalign | kalign | sequence aligner | http://msa.sbc.su.se/cgi-bin/msa.cgi |
| trimAl | trimal | sequence trimmer | http://trimal.cgenomics.org/ |
| Gblocks | Gblocks | sequence trimmer | http://molevol.cmima.csic.es/castresana/Gblocks.html |
| FastTree | fasttree | tree reconstructor | http://www.microbesonline.org/fasttree/ |

See the section [3.8](#38-adding-third-party-software-in-the-pipeline) to find out more details on how to configure a third-party software to be part of TaxOnTree phylogenetic pipeline.

> **Note**: The requirement of software of each type depends on the input provided by the user. For instance, if an aligned Multi-FASTA sequence is provided as input, TaxOnTree will require only a software for tree reconstruction. In the other hand, if a single accession from NCBI is provided, TaxOnTree will require software of all four types to perform the analysis.

### 3.2. Installation:

#### 3.2.1. Downloading TaxOnTree

Use the following commands to download TaxOnTree in a local machine:

```bash
git clone 
cd taxontree
```
#### 3.2.2. Configuring your email address

Set your email address to allow TaxOnTree to perform HTTP request to NCBI and/or Uniprot servers. For this, edit the file CONFIG.xml that accompanies this package and set your email address between <email> tags.

```
<email>your email goes here</email>
```

#### 3.2.3. Testing TaxOnTree

Finally, test TaxOnTree by running the following command:

```
./taxontree
```

If all goes well, it should display the following message.

```
        TaxOnTree  Copyright (C) 2015-2017  Tetsu Sakamoto
        This program comes with ABSOLUTELY NO WARRANTY.
        This is free software, and you are welcome to redistribute it under
        certain conditions. See GNU general public license v.3 for details.

ERROR: No input was provided.
Usage:
    ./taxontree -singleID <sequence_ID>

    ./taxontree -seqFile <FASTA_file>

    ./taxontree -listFile <list_file>

    ./taxontree -treeFile <tree_file> -queryID <query_id>

    Inputs:
        [-seqFile FASTA_file] [-listFile list_file] [-singleID sequence_ID]
        [-treeFile tree_file] [-blastFile blast_file] [-mfastaFile
        mfasta_file] [-alignFile align_file]

    Blast options:
        [-db database_name] [-evalue] [-threshold] [-maxTarget int_value]
        [-maxTargetBlast int_value]

    Alignment options:
        [-aligner] [-trimming]

    Tree options:
        [-treeProg] [-treeTable table_file] [-printLeaves] [-treeRoot]
        [-leafFmt]

    Filter options:
        [-showIsoform] [lcaLimit] [taxFilter] [taxFilterCat] [restrictTax]

    Other parameters:
        [-out file_name] [-queryID query_id] [-queryTax tax_id] [-txidMap
        tax_id] [-position] [-delimiter] [-numThreads] [-mysql]
        [-taxRepFormat] [-forceNoTxid] [-version]

    Help:
        [-help] [-man]

        Use -man for a detailed help.
```

TaxOnTree is ready to use  in  most  of  Unix  Platform. But it only works if the folders lib/ that follow this script are  in  the same location of execution. If you want to freely run TaxOnTree  in  other  location,  add the TaxOnTree folder into the environment variable  by  using,  for 	example, the following commands:
	
 ```bash
	> echo "export PATH=$PATH:/path/to/program/taxontree/" >> ~/.bash_profile
	> source ~/.bash_profile
```


### 3.3. Inputs

#### 3.3.1. Single ID (-singleid <single_id>)

Single protein accession number from NCBI (GI or accession number) or UniprotKB (accession number or entry name). Example: "P04156" or "4757876" or "PRIO_HUMAN".

#### 3.3.2. Sequence file (-seqFile <fasta_file>)

A file containing a single sequence in FASTA format. You may provide the taxonomy ID of your sequence
by using the parameter -queryTax <taxonomy_id> in the command line, or else, TaxOnTree will attribute
the taxonomy ID of the Blast best hit as its taxonomy ID.

Example:
```
>Human_Prion
MANLGCWMLVLFVATWSDLGLCKKRPKPGGWNTGGSRYPGQGSPGGNRYPPQGGGGWGQP
HGGGWGQPHGGGWGQPHGGGWGQPHGGGWGQGGGTHSQWNKPSKPKTNMKHMAGAAAAGA
VVGGLGGYMLGSAMSRPIIHFGSDYEDRYYRENMHRYPNQVYYRPMDEYSNQNNFVHDCV
NITIKQHTVTTTTKGENFTETDVKMMERVVEQMCITQYERESQAYYQRGSSMVLFSSPPV
ILLISFLIFLIVG
```

#### 3.3.3. Blast file (-blastFile <blast_file>)

Blast result of a single protein generated by Standalone BLAST+ in tabular format (-outfmt 6). You may provide a protein accession 
from the result to be considered as query using -queryID, or else, TaxOnTree will consider the protein 
in the first column or, if it is not an identifier from NCBI or Uniprot, the best hit subject as query.

#### 3.3.4. List file (-listFile <list_file>) 

A file containing a list of protein identifiers from NCBI (GI or accession number) or UniprotKB 
(accession number or entry name) separated by new line. You may provide a protein accession from 
the list to be considered as query using -queryID, or else, TaxOnTree will consider the first entry 
in the list as query. 

#### 3.3.5. Multi-FASTA file (-mfastaFile <mfasta_file>)

A Multi-FASTA file containing ortholog sequences. You may provide a protein accession from the file to be 
considered as query using -queryID, or else, TaxOnTree will consider the first entry in the file as query.

#### 3.3.6. Aligned multi-FASTA file (-alignFile <align_file>)

An Aligned Multi-FASTA file. You may provide a protein accession from the list to be considered as 
query using -queryID, or else, TaxOnTree will consider the first entry in the file as query.

#### 3.3.7. Tree file (-treeFile <tree_file>)

A tree file in NEWICK format. You must provide the protein in the tree to be considered as query by
using -queryID. If the leaves of the tree are not exactly an accession number from NCBI or UniprotKB, but it contains the accessions in their names, you can use -delimiter and -position parameters to allow TaxOnTree to extract the accession and retrieve the taxonomy ID of each proteins. Alternatively, you can provide a table containing the leaves (first column) and their
corresponding taxonomy id (second column) using the option -treetable. To generate a file with all leaves
name in your tree, use the option -printLeaves. 

### 3.4. Parameters

#### 3.4.1. Blast option

* **-db <database_name> Default: refseq_protein**

BLAST-formatted database name. It works on -seqFile, -singleID, -blastFile and -listFile.

For Standalone protein BLAST search, provide the address and the name (without extension) of the database. 
Example: /home/user/taxontree/db, where db is the name of BLAST-formatted database. This option will
only work if the database was generated using sequences from GenBank or Uniprot (using its 
FASTA-header pattern) and the option -parse_seqids on makeblastdb command (See details on README).

To request the BLAST search from NCBI server, you may choose one of these databases: 
nr or refseq_protein. 

* **-evalue <real_value> Default: 10e-5**

Expect value threshold for BLAST search. It works on -seqFile and -singleID.

* **-threshold <int_value> Default: 50**

Protein identity threshold. For each subject, TaxOnTree calculates its identity with the query sequence after
removing overlapping HSPs and considering the length of the query sequence. Threshold may vary between 0-100.
It works on -seqFile, -blastFile and -singleID.

* **-maxTarget <int_value> Default: 200**

Max target sequences to be used for phylogenetic analysis. It works on -seqFile, -blastFile and -singleID.

* **-maxTargetBlast <int_value>**

Max target sequence to be retrieved by BLAST. It works on -seqFile and -singleID.

#### 3.4.2 Alignment option

* **-aligner <aligner_software> Default: muscle

Software for sequence alignment step. It works on -seqFile, -blastFile, -listFile, -mfastaFile and -singleID. 
To add more aligners, see CONFIG.xml.

* **-trimming <trimming_software> Default: trimal

Software for alignment trimming step. You can set "false" to skip this step. It works on -seqFile, -blastFile, -listFile, -mfastaFile, 
-alignFile and -singleID. To add more alignment trimming software, see CONFIG.xml.

#### 3.4.3. Tree option

* **-treeProg <tree_reconstruction_software> Default: FastTree

Software for tree reconstruction. It works on -blastFile, -listFile, -mfastaFile, -alignFile and -singleID. To add more tree 
reconstruction software, see CONFIG.xml.

* **-treeTable <table_file>

A table containing the leaf names of the input tree file in the first column and the correspondent taxonomy
ID in the second. Use this option if the input tree does not contain an accession from NCBI or Uniprot
in its leaves. To obtain a list of leaf names in your tree, use the option -printLeaves. It works on -treeFile.

Example: For a newick tree "(gorilla,(human,chimp))", you could provide a tab-delimited table like this:

	human	9606
	chimp	9598
	gorilla	9595

* **-printLeaves

Print the leaf names comprising your tree and exit. Use this to help you making the tree table file. It works on -treeFile.

* **-treeRoot Default: 1

Define tree rooting mode. Use 0 to skip this step, 1 to root at midpoint or 2 to use taxonomic information do define a root. It works on all input.

* **-leafFmt Default: "lcaN;id;geneName;species;rankcode(family,order,class)"

Leaf name format displayed in the tree. It works on all input. Data available to display in the leaf are: lcaN, lca, id, accession, species, geneID, geneName, rankcode, rankname. Use semicolons to separate the different data type. 

For "rankcode" and "rankname", include the taxonomic ranks that you want to display separated with comma and delimited by parenthesis. Taxonomic rank options: superkingdom, kingdom, phylum, subphylum, superclass, class, subclass, superorder, order, suborder, superfamily, family, subfamily, genus, subgenus, species, subspecies. 

#### 3.4.4 Filter option

* **-showIsoform

TaxOnTree automatically links the RefSeq or Uniprot protein to a GeneID and discards its isoforms 
from further analysis. Use this option to allow isoforms in the tree. It works on all inputs.

* **-lcaLimit <int_value>

Exclude all sequences from organisms which the LCA with the query organism is below the provided level. 
It works on all inputs.

* **-taxFilterCat

Filter sequences by category which could be a taxonomic rank or LCA. If "kingdom" is provided, it'll leave 
sequences from N organisms in each kingdom found in the tree. Use -taxFilter to define the N. It works on all inputs.

Categories allowed: lca, superkingdom, kingdom, phylum, subphylum, superclass, class, subclass, superorder, order, suborder, 
superfamily, family, subfamily, genus, subgenus, species, subspecies.

* **-taxFilter <int_value>

Filter sequences by category which could be a taxonomic rank or LCA. If 2 is provided, it'll leave 
sequences from 2 organisms in each category found in the tree. Use -taxFilterCat to define the category. 
It works on all inputs.

* **-restrictTax <list_file>

Provide a list of taxonomy ID (separated by newline) to show only sequences belonging to organisms which 
have their taxonomy ID listed in the file. It works on all inputs.

#### 3.4.5. Other parameters

* **-txidMap <taxonomy_id>

Force TaxOnTree to consider the taxonomy ID provided by this option to be mapped in the tree. Example: "9606"
for human. It works on all inputs. 

* **-queryTax <taxonomy_id>

NCBI taxonomy ID assigned to the query protein. Example: "9606" for human. It works only in -seqFile. 

* **-queryID <query_id>

Protein accession or name in the list or tree to be considered as query. It works on -blastFile, -listFile, -mfastaFile
-alignFile and -treeFile. 

* **-forceNoTxid

Retain entries in which TaxOnTree could not determine their taxonomy ID. These entries will
be assigned with the taxonomy ID of root (txid:1). It works on all inputs.

* **-forceNoInternet

Force TaxOnTree to not retrieve data from internet. It works on all inputs.

* **-out <file_name> Default: Input name

Prefix for output files. It works on all inputs.

* **-mysql

Use a local MySQL database to retrieve required TaxOnTree data (See README for details). It works on all inputs.

* **-numThreads Defalut: 1

Number of processors to be used for programs that can handle multi-threading. It works on all inputs.

* **-version

Print TaxOnTree version.

### 3.5. Outputs



### 3.6. Configuring a local MySQL database

### 3.7. Making your own sequence database

### 3.8. Adding third-party software in the pipeline

## 4. Opening and exploring the Nexus tree on FigTree

After running TaxOnTree from [web interface](http://biodados.icb.ufmg.br/taxontree/) or from [command line](https://sourceforge.net/projects/taxontree/), TaxOnTree will generate a **Nexus file** structured to be opened in [FigTree](http://tree.bio.ed.ac.uk/software/figtree/) software. FigTree is a free graphical viewer of phylogenetic trees developed in Java by Andrew Rambaut group. There are versions for Mac, UNIX and Windows. So, once you have the Nexus file, just download FigTree and open the Nexus file in it. In this manual, we will use a [tree sample]() available in TaxOnTree website.

### 4.1. Exploring the taxonomic relationship by LCA

After opening the Nexus file in FigTree, you will see your phylogenetic tree with the branches colored according to the LCA (Lowest Common Ancestor) between the query species (in red) and the other species in the tree (**Figure S3**). LCA of two or more organisms represents the most recent ancestor that all organisms in the set have in common (see **Box 1** for mor details). There will have also a legend for the colors used in the tree. 

> **Note**: Some taxon names in the legend are followed by an asterisk. This indicates that there is no organism in the tree that has this taxon as the LCA with the query organism.

### 4.2. Exploring the taxonomic diversity by ranks

Did you ever asked yourself how many distinct class, order or family are in your tree? You can answer this in a few seconds with a tree generated by TaxOnTree in hand. TaxOnTree also embeds in the tree the taxonomic lineage data of all taxa comprising the phylogenetic tree, allowing you to colorize the tree according to a taxonomic rank. Take the following steps in the FigTree software and, if you are curious about how missing ranks in some lineages are filled in TaxOnTree, check the **Box 2**.

1. In the FigTree side menu, go to *Appearance*;
1.	Click on *Colour by*, and choose a taxonomic rank to be used to color the branches. By choosing *12-family*, for example (**Figure S4**), the tree will be colored according to the family rank. You can setup the branch colors by clicking in *Colours*;
1.	Branch color will change immediately. To make the color legend concordant to the branch color, go to *Legend* in the FigTree side menu and, in *Attribute*, set the same taxonomic rank that you selected in the previous step (**Figure S5**).

### 4.3. Adding or changing labels in the tree

Taxonomic data can also be evidenced on tip, node and/or branch labels in the tree. For instance, to change the tip label, go to *Tip labels*, on FigTree side menu; and on *Display* select a taxonomic rank to be diplayed at the tip of the tree.

To evidence the taxonomic data on the nodes and/or on branches of the tree, use the same procedure but at *Node label* and/or *Branch label*, respectively.

### 4.4. Checking branch statistic support value

By default, the statistic support of a branch are evidenced by the size of the nodes, i.e. The bigger the node size, the higher the support. If you want to know their values go to *Node label* on FigTree side menu and choose *BOOT* on *Display*.

### 4.5. Evidencing duplication nodes

TaxOnTree also annotates those nodes that represent duplication event. To evidence by node size, on FigTree side menu, go to *Node shapes* and choose *dup* on *Size by*. Set also an appropriate value for *Max size* and *Min size* parameters.

## 5. Sample tree images

## 6. Contact

If you have troubles or suggestions to improve our work, please contact us by the following email address:

* tetsufmbio@gmail.com (Tetsu Sakamoto)
* miguel@ufmg.br (J. Miguel Ortega)

_**Laboratório de Biodados**  
Instituto de Ciências Biológicas (ICB)  
Universidade Federal de Minas Gerais (UFMG)  
Belo Horizonte - Minas Gerais – BRAZIL  
Zip code: 31270-010_
